allennlp/models/srl_bert.py

```diff
@@ -136,6 +137,17 @@ def decode(self, output_dict: Dict[str, torch.Tensor]) -> Dict[str, torch.Tensor
         Does constrained viterbi decoding on class probabilities output in :func:`forward`.  The
         constraint simply specifies that the output tags must be a valid BIO sequence.  We add a
         ``"tags"`` key to the dictionary with the result.
+
+        NOTE: First, we decode a BIO sequence on top of the wordpieces. This is important; viterbi
+        decoding produces low quality output if you decode on top of word representations directly,
+        because the model already learns very strong preferences for the BIO tag type.
+
+        Secondly, it's important that the indices we use to recover words from the wordpieces are the
+        start_offsets (i.e offsets which correspond to using the first wordpiece of words which are
+        tokenized into multiple wordpieces) as otherwise, we might get an ill-formed BIO sequence
```

Comment:
I'm a bit confused about what's going on here, probably because I'm lacking context on how word tags are split into wordpiece tags in the first place.  If you get a bunch of `I-` tags from taking the last wordpiece, wouldn't you get a bunch of `B-` tags from taking the first one?  Or is it that all continuation wordpieces always have `I-`, but the first wordpiece will have the original word tag?  Then this makes more sense.

Unless you're using BIOUL, then this still gets messed up, because you might have `L-`, `I-`...

Reply:
Yes, it's basically that we need the `B-XXX` tag and not the `I-XXX` tag for words that have been split into multiple wordpieces.